File: Src/java/engine/src/main/java/org/opencds/cqf/cql/engine/elm/execution/DateFromEvaluator.java
package org.opencds.cqf.cql.engine.elm.execution;

import org.opencds.cqf.cql.engine.exception.InvalidOperatorArgument;
import org.opencds.cqf.cql.engine.execution.Context;
import org.opencds.cqf.cql.engine.runtime.Date;
import org.opencds.cqf.cql.engine.runtime.DateTime;
import org.opencds.cqf.cql.engine.runtime.Precision;

/*
date from(argument DateTime) Date

NOTE: this is within the purview of DateTimeComponentFrom
  Description available in that class
*/

public class DateFromEvaluator extends org.cqframework.cql.elm.execution.DateFrom {

    public static Date dateFrom(Object operand) {
        if (operand == null) {
            return null;
        }

        if (operand instanceof DateTime) {
            if (((DateTime) operand).getPrecision().toDateTimeIndex() < 1) {
                return (Date) new Date(((DateTime) operand).getDateTime().getYear(), 1, 1).setPrecision(Precision.YEAR);
            }
            else if (((DateTime) operand).getPrecision().toDateTimeIndex() < 2) {
                return (Date) new Date(((DateTime) operand).getDateTime().getYear(), ((DateTime) operand).getDateTime().getMonthValue(), 1).setPrecision(Precision.MONTH);
            }
            else {
                return (Date) new Date(((DateTime) operand).getDateTime().getYear(), ((DateTime) operand).getDateTime().getMonthValue(), ((DateTime) operand).getDateTime().getDayOfMonth()).setPrecision(Precision.DAY);
            }
        }

        throw new InvalidOperatorArgument("date from(DateTime)", String.format("date from(%s)", operand.getClass().getName()));
    }

    @Override
    protected Object internalEvaluate(Context context) {
        Object operand = getOperand().evaluate(context);
        return dateFrom(operand);
    }
}

File: Src/java/engine/src/commonMain/kotlin/org/opencds/cqf/cql/engine/elm/executing/DateFromEvaluator.kt
package org.opencds.cqf.cql.engine.elm.executing

import kotlin.jvm.JvmStatic
import org.opencds.cqf.cql.engine.exception.InvalidOperatorArgument
import org.opencds.cqf.cql.engine.runtime.Date
import org.opencds.cqf.cql.engine.runtime.DateTime
import org.opencds.cqf.cql.engine.runtime.Precision
import org.opencds.cqf.cql.engine.util.javaClassName

/*
date from(argument DateTime) Date

NOTE: this is within the purview of DateTimeComponentFrom
  Description available in that class
*/
object DateFromEvaluator {
    @JvmStatic
    fun dateFrom(operand: Any?): Date? {
        if (operand == null) {
            return null
        }

        if (operand is DateTime) {
            if (operand.precision!!.toDateTimeIndex() < 1) {
                return Date(operand.dateTime!!.getYear(), 1, 1).withPrecision(Precision.YEAR)
                    as Date
            } else if (operand.precision!!.toDateTimeIndex() < 2) {
                return Date(operand.dateTime!!.getYear(), operand.dateTime!!.getMonthValue(), 1)
                    .withPrecision(Precision.MONTH) as Date
            } else {
                return Date(
                        operand.dateTime!!.getYear(),
                        operand.dateTime!!.getMonthValue(),
                        operand.dateTime!!.getDayOfMonth(),
                    )
                    .withPrecision(Precision.DAY) as Date
            }
        }

        throw InvalidOperatorArgument("date from(DateTime)", "date from(${operand.javaClassName})")
    }
}
